predictionnet/validator/reward.py

```diff
@@ -107,42 +196,61 @@ def get_rewards(
                                     seconds=timestamp.second,
-                                    microseconds=timestamp.microsecond) + timedelta(minutes=INTERVAL + 5, seconds=30)
+                                    microseconds=timestamp.microsecond) + timedelta(minutes=prediction_interval, seconds=1)
     
     ny_timezone = timezone('America/New_York')
 
-    while (datetime.now(ny_timezone) < rounded_up_time - timedelta(minutes=4, seconds=30)):
-        bt.logging.info(f"Waiting for next {INTERVAL}m interval...")
-        if(datetime.now(ny_timezone).minute%10==0):
+    while (datetime.now(ny_timezone) < rounded_up_time):
+        bt.logging.info(f"Waiting for next {prediction_interval}m interval...")
+        if(datetime.now(ny_timezone).minute%20==0):
             self.resync_metagraph()
         time.sleep(15)
 
-    current_time_adjusted = rounded_up_time - timedelta(minutes=INTERVAL + 5)
-    print(rounded_up_time, rounded_up_time.hour, rounded_up_time.minute, current_time_adjusted)
     
-    data = yf.download(tickers=ticker_symbol, period='1d', interval='5m')
-    bt.logging.info("Procured data from yahoo finance.")
-
-    bt.logging.info(data.iloc[-7:-1])
-    close_price = data['Close'].iloc[-7:-1].tolist()
+    data = yf.download(tickers=ticker_symbol, period='5d', interval='5m', progress=False)
+    #bt.logging.info("Procured data from yahoo finance.")
+    # add an extra timepoint for dir_acc calculation
+    bt.logging.info(data.iloc[(-N_TIMEPOINTS-1):])
+    close_price = data['Close'].iloc[(-N_TIMEPOINTS-1):].tolist()
     close_price_revealed = ' '.join(str(price) for price in close_price)
 
     bt.logging.info(f"Revealing close prices for this interval: {close_price_revealed}")
 
-    # Get all the reward results by iteratively calling your reward() function.
-    scoring = [reward(response, close_price) if response.prediction != None else 0 for response in responses]
-    worst_loss = max(scoring)
-    bt.logging.debug(worst_loss)
-    scoring = [1 - (score / (worst_loss + 5)) if score != 0 else 0 for score in scoring] #constant 5 is added so that models with worst loss don't get 0 rewards
-    return torch.FloatTensor(scoring)
-
-    #scaler = MinMaxScaler(feature_range=(0,1))
-    #return torch.FloatTensor(scaler.fit_transform(np.array(scoring).reshape(-1, 1)).flatten()).to(self.device)
+    # Preallocate an array (nMiners x N_TIMEPOINTS x N_TIMEPOINTS) where the third dimension is t-1, t-2,...,t-N_TIMEPOINTS for past predictions
+    raw_deltas = np.full((len(responses),N_TIMEPOINTS,N_TIMEPOINTS), np.nan)
+    raw_correct_dir = np.full((len(responses),N_TIMEPOINTS,N_TIMEPOINTS), False)
+    ranks = np.full((len(responses),N_TIMEPOINTS,N_TIMEPOINTS), np.nan)
+    for x,response in enumerate(responses):
+        # calc_raw also does many helpful things like shifting epoch to 
+        delta , correct = calc_raw(self, miner_uids[x], response, close_price)
```

Comment:
I think there is a problem with close_price alignment. Maybe this is originated from modified base_miner code for testing the new reward system. modified base_miner responds with past close prices not the future close prices.
- The current implementation only waits 5 minutes before fetching close prices, leading to a misalignment between the prediction intervals and the actual close prices fetched. Specifically:
  - **Predictions**: 5, 10, 15, 20, 25, 30 minutes into the future from the prediction timestamp.
  - **Fetched Close Prices**: -25, -20, -15, -10, -5, 0, 5 minutes from the prediction timestamp.
- This misalignment results in inaccurate calculations within the `calc_raw` function:
  - The `calc_raw` function requires accurate close prices that match the prediction intervals (i.e., 5, 10, 15, 20, 25, and 30 minutes after the prediction timestamp).

Reply:
There is no misalignment. The reward function is called first, so the synapse, as is when it enters calc_raw, has one prediction that has already matured (response.prediction[0]). So it is actually:

If the reward function is called just after 11 am:
- **Predictions:** 0, 5, 10, 15, 20, 25
    - this prediction is from 5 minutes ago (requested at 10:55):
    - [11:00, 11:05, 11:10, 11:15, 11:20, 11:25]
- **Fetched close prices:** -25, -20, -15, -10, -5, 0
    - [10:35, 10:40, 10:45, 10:50, 10:55, 11:00] 

Then the time_shift function removes all predictions that haven't matured yet. This results in the matricies used to calculate delta and correct_dirs looking like this:

- **Predictions:**
- [[nan, nan, nan, nan, nan, 0], = 5min ago prediction
    [nan, nan, nan, nan,  -5,  0], = 10min ago prediction
    [nan, nan, nan,  -10,  -5,  0],      ...
    [nan, nan,  -15,  -10,  -5,  0],
    [nan,  -20, -15,  -10,  -5,  0],
    [-25,  -20, -15,  -10,  -5,  0]] = 30 min ago prediction
- each column is now alligned and refers to a specific time: [10:35, 10:40, 10:45, 10:50, 10:55, 11:00], which is mirrored by the fetched close prices array. 
- **Fetched close prices:**
- [[-25,  -20, -15,  -10,  -5,  0], = 5min ago prediction
    [-25,  -20, -15,  -10,  -5,  0], = 10min ago prediction
    [-25,  -20, -15,  -10,  -5,  0],      ...
    [-25,  -20, -15,  -10,  -5,  0],
    [-25,  -20, -15,  -10,  -5,  0],
    [-25,  -20, -15,  -10,  -5,  0]] = 30 min ago prediction

Reply:
with new commits, I can see that deltas are printed correctly. But I think direction results are weird.
I call get_reward function with these predictions and close prices. The close price when first prediction requested is 1000.
- pred: [1005, 1010, 1015, 1020, 1025, 1030], close price: [970, 976, 982, 988, 994, 1000, 1006]
- pred: [1010 ,1015, 1020, 1025, 1030, 1035], close price: [976, 982, 988, 994, 1000, 1006, 1012]
- pred: [1015, 1020, 1025, 1030, 1035, 1040], close price: [982, 988, 994, 1000, 1006, 1012, 1018]
- pred: [1020, 1025, 1030, 1035, 1040, 1045], close price: [988, 994, 1000, 1006, 1012, 1018, 1024]
- pred: [1025, 1030, 1035, 1040, 1045, 1050] close price: [994, 1000, 1006, 1012, 1018, 1024, 1030]
- pred: [1030, 1035, 1040 ,1045, 1050, 1055], close price: [1000, 1006, 1012, 1018, 1024, 1030, 1036]

The last result of deltas, correct_dirs of calc_raw function is below.
Delta is perfect, but correct_dir is weird. I think all rows except the except the first row should be 1. The direciton of predictions are good, as I know.
```
delta
[[nan nan nan nan nan  6.]
 [nan nan nan nan  5.  6.]
 [nan nan nan  4.  5.  6.]
 [nan nan  3.  4.  5.  6.]
 [nan  2.  3.  4.  5.  6.]
 [ 1.  2.  3.  4.  5.  6.]]
correct_dir
[[nan nan nan nan nan  0.]
 [nan nan nan nan  0.  0.]
 [nan nan nan  0.  0.  0.]
 [nan nan  0.  0.  1.  1.]
 [nan  0.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.  1.]]
```

Reply:
good catch. Looks like the reference point for the prediction array was earlier prediction array points, instead of close_price points. Also to clarify, there should be one close price vector (repeated 6 times) per 6 prediction vectors. So in your example, I believe close_price would be [970, 976, 982, 988, 994, 1000, 1006] stacked 6 times into a matrix.

Reply:
with latest commit, there is still something wrong with correct_dirs result of calc_raw function.
This is the result of last call of calc_raw function with same input as above comment.
pred_dir, close_dir, delta is understandable, but I can't understand correct_dir data.
```
pred_dir
[[  0.  -5. -10. -15. -20. -25.]
 [ -1.  -6. -11. -16. -21. -26.]
 [ -2.  -7. -12. -17. -22. -27.]
 [ -3.  -8. -13. -18. -23. -28.]
 [ -4.  -9. -14. -19. -24. -29.]
 [ -5. -10. -15. -20. -25. -30.]]

close_dir
[[ 24  18  12   6   0  -6]
 [ 18  12   6   0  -6 -12]
 [ 12   6   0  -6 -12 -18]
 [  6   0  -6 -12 -18 -24]
 [  0  -6 -12 -18 -24 -30]
 [ -6 -12 -18 -24 -30 -36]]

delta
[[nan nan nan nan nan  6.]
 [nan nan nan nan  5.  6.]
 [nan nan nan  4.  5.  6.]
 [nan nan  3.  4.  5.  6.]
 [nan  2.  3.  4.  5.  6.]
 [ 1.  2.  3.  4.  5.  6.]]

correct_dir
[[nan nan nan nan nan  1.]
 [nan nan nan nan  0.  0.]
 [nan nan nan  0.  0.  0.]
 [nan nan  0.  0.  1.  1.]
 [nan  0.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.  1.]]
```

Reply:
> good catch. Looks like the reference point for the prediction array was earlier prediction array points, instead of close_price points. Also to clarify, there should be one close price vector (repeated 6 times) per 6 prediction vectors. So in your example, I believe close_price would be [970, 976, 982, 988, 994, 1000, 1006] stacked 6 times into a matrix.

I wanted to test the reward system over time for a single miner, not to check the reward of multiple miners for a single prediction. In my example data, one row represents the prediction and close price for each timepoint. So, 6 rows means 30 minutes of data with prediction and close price every 5 minutes. That's why the close_price is appended with a new data each time.

Reply:
How I described it is for a single miner. Each miner is evaluated on all predictions made up to 30 minutes ago, so a single set of 6 close prices, and that set of prices is compared to all 6 prediction epochs at various maturity levels. That said, time_shift was being called on close_price when it shouldn't be so I fixed that.

Reply:
Got it. Now the correct_dir result is printed correctly as I understand.